File: leetcode_solutions/valid-parentheses/valid-parantheses.java
import java.util.Stack;
class Solution {
  public boolean isValid(String s) {
    Stack<Character> stack = new Stack<Character>(); // declaring a stack which accepts Character (generic)
    for (char c : s.toCharArray()) { // transforming our string to a character array then using an iterator to traverse each element
      if (c == '[' || c == '(' || c == '{') { // checking to if we have an opening character if yes then we push it onto the stack
        stack.push(c);
      } else if (c == ')' && !stack.isEmpty() && stack.peek() == '(') { // checking for a closing bracket, if c is one and the next element in the stack is an opening one we pop
        stack.pop();
      } else if (c == ']' && !stack.isEmpty() && stack.peek() == '[') {
        stack.pop();
      } else if (c == '}' && !stack.isEmpty() && stack.peek() == '{') {
        stack.pop();
      } else {
        return false;
      }
    }
    return stack.isEmpty(); // if the stack isEmpty then we have checked everything successfully.
  }
}

File: leetcode_solutions/valid-parentheses/valid-parantheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        if len(s) % 2 != 0: 
            return False
        for c in s: 
            if c == '(' or c =='[' or c =='{':
                stack.append(c)
            elif (c == ']' and len(stack) > 0 and stack.pop() != '['):
                return False
            elif (c == ')' and len(stack) > 0 and stack.pop() != '(' ):
                return False
            elif (c == '}' and len(stack) > 0 and stack.pop() != '{' ):
                return False
        return len(stack) == 0
    
    # make a variable i = stack.pop to avoid unexpected behavior
